pyflakes/checker.py

```diff
@@ -616,9 +628,13 @@ def handleNode(self, node, parent):
             node.col_offset += self.offset[1]
         if self.traceTree:
             print('  ' * self.nodeDepth + node.__class__.__name__)
-        if self.futuresAllowed and not (isinstance(node, ast.ImportFrom) or
-                                        self.isDocstring(node)):
-            self.futuresAllowed = False
+
+        if (isinstance(self.scope, ModuleScope) and
+                self.scope._futures_allowed and
```

Comment:
would adding `_futures_allowed = False` to the `Scope` base class allow the `isinstance(self.scope, ModuleScope)` to be dropped from this logic?